doc/explanation/mental-model.rst

```diff
@@ -0,0 +1,175 @@
+The Dune Mental Model
+=====================
+
+It is not strictly necessary to understand Dune's underlying model to use it;
+but knowing how it works under the hood will help understand some errors and
+what's possible with Dune.
+
+.. note::
+
+   This document is a simplification of the reality: the actual rules might be
+   different, it does not touch rule loading and glosses over how caching
+   works, but should be a useful tool to build an understanding of Dune.
+
+How Dune Works
+--------------
+
+The building brick of Dune is the *rule*:
+
+    A *rule* reads *dependencies* and writes *targets* by executing an *action*.
+
+When ``dune build`` is executed, it will first read ``dune`` files to determine
+the rules that apply to the project. Once it has done this, it will determine
+what actions it needs to execute to build the required targets.
+
+An Example
+----------
+
+Let's take the following example.
+
+- there's a CLI tool written in OCaml.
+- it has some build-time configuration stored in ``config.json``.
+- it has an integration test which consists in running in it with
+  ``testdata.txt`` as input.
+
+Configuration Generation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+To express the generation of the configuration module we could write:
+
+.. code:: dune
+
+  (rule
+   (deps convert/json2ml.exe config.json)
+   (target config.ml)
+   (action
+    (run convert/json2ml.exe config.json -o config.ml)))
+
+This rule will:
+
+- read its dependencies: ``convert/json2ml.exe`` and ``config.json``
+- and write its target: ``config.ml``
+- using an action: ``(run convert/json2ml.exe config.json -o config.ml)``
+
+This rule is very explicit: we write a stanza for a single Dune rule.
+
+Building the Executable
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In contrast, to describe the compilation of the executable, we would write:
+
+.. code:: dune
+
+  (executable
+   (name tool)
+   (modules main config))
+
+Here, we use Dune's abstractions. It knows about the OCaml compilation model:
+the modules need to be compiled and linked together. So it will generate the
+following rules under the hood:
+
+- one rule to compile the ``Main`` module:
+
+  - it will read its dependency: ``main.ml``
+  - and write its output: ``main.cmx``
+  - using an action: ``(run ocamlopt -c main.ml)``
+
+- one rule to compile the ``Config`` module:
+
+  - it will read its dependency: ``config.ml``
+  - and write its output: ``config.cmx``
+  - using an action: ``(run ocamlopt -c config.ml)``
+
+- one rule to compile the ``tool.exe`` executable:
+
+  - it will read its dependencies: ``main.cmx`` and ``config.cmx``
+  - and write its output: ``tool.exe``
+  - using an action: ``(run ocamlopt -o tool.exe main.cmx config.cmx``)
+
+Note that in this example, some files are targets of a rule and dependencies of
+another (``.cmx`` files). We are unlikely to ever interact with them directly,
+so it can also be useful to think of the ``(executable)`` stanza as acting as a
```

Comment:
I'm curious why you find it useful to think of this as just one rule. From the point of view of Dune's performance/cache model, it's pretty important that these rules are separate. (On the other hand, bundling these separate rules into one is sometimes appropriate and internally we have some optimisations that are actually working this way.)

Reply:
As a second opinion, I find it useful to think about it that way, since as a Dune user one rarely if ever need to think about `.cm*` files, so the intermediate rules aren't all that interesting.

I would even consider making this thinking the starting point since that's sort of the intuition and then explain the workings in more detail later.

Reply:
This comes from this graph from the presentation (omitting the graphs in a first version is a deliberate design decision, so that the text has to be clear and stays accessible)

```mermaid
graph LR
  tool.exe;
  config.ml;
  main.ml;
  subgraph "executable: tool"
    bm[/compile Main/] --> main.cmx --> l[/link executable/]
    bc[/compile Config/] --> config.cmx
  end
  
  l[/link executable/] --> tool.exe
  config.cmx --> l
  config.ml --> bc
  main.ml --> bm
```

Here what I meant is that the collection of rules from the executable stanza act as an opaque box with the `ml` files as dependencies and `tool.exe` as a target. I think that's useful to think it that way but I shouldn't call that a rule.

Reply:
I like the graph! My feeling is that it's sufficient to say that the highlighted group of rules corresponds to the `executable` stanza and leave it at that (i.e. no need to talk about thinking about them as one rule).